src/java/org/apache/cassandra/db/compaction/unified/AdaptiveController.java

```diff
@@ -135,19 +145,82 @@ static Controller fromOptions(Environment env,
                                   Map<String, String> options)
     {
         int scalingParameter = options.containsKey(STARTING_SCALING_PARAMETER) ? Integer.parseInt(options.get(STARTING_SCALING_PARAMETER)) : DEFAULT_STARTING_SCALING_PARAMETER;
-        int[] scalingParameters = new int[32];
-        int[] previousScalingParameters = new int[32];
-        //set the scaling parameter for each level to the starting scaling parameter or default scaling parameter
-        Arrays.fill(scalingParameters, scalingParameter);
-        Arrays.fill(previousScalingParameters, scalingParameter);
+        int[] scalingParameters;
+        int[] previousScalingParameters;
+        long currentFlushSize = flushSizeOverrideMB << 20;
+
+        File f = getControllerConfigPath();
+        try
+        {
+            JSONParser jsonParser = new JSONParser();
+            JSONObject jsonObject = (JSONObject) jsonParser.parse(new FileReader(f));
+            scalingParameters = readStoredScalingParameters((JSONArray) jsonObject.get("scaling_parameters"));
+            if (scalingParameters != null)
+            {
+                logger.debug("Successfully read stored scaling parameters from disk.");
+                previousScalingParameters = scalingParameters.clone();
+            }
+            else
+            {
+                logger.warn("Unable to read scaling_parameters. Using starting value instead.");
+                scalingParameters = new int[UnifiedCompactionStrategy.MAX_LEVELS];
```

Comment:
This pattern repeats several times (here and in the catch clauses).

In all cases the values for `scalingParameters` are copied into `previousScalingParameters`, so we can move that processing after the `try...catch`. Also, why not start with null for `scalingParameters`, and check if it is still null after the `try...catch` block?